### PR TITLE
Fully variable nonbonded lambda interpolation

### PR DESCRIPTION
**Pull request by proteneer**
This PR implements a JIT-based compiler using regex substitution similar to OpenMM for allowing the end-user to specify arbitrary transformations of lambda on the charges, sigmas, epsilons, and w component. In particular, this is accomplished by enabling four extra strings on the constructor of the Nonbonded class, specifying the transformations that should be applied to the lambda parameter.

It roughly follows the design in #402, namely that parameters are interpolated via some arbitrary transformation of the form:
```
p(lambda) = (1-f(lambda))*p_src + f(lambda)*p_dst, s.t. f(0)=0 and f(1)=1
```
The 4D distance/softcore similary interpolates between 0 and 1*cutoff.

The neatest part of this PR is that the derivatives for `f(lambda)` is implemented using the complex step derivative. Which is numerically accurate to  machine precision. The set of operators we allow is slightly less than that of lepton from OpenMM, but this has the major advantage of *not* requiring any symbolic analysis, as the complex step AD is nearly identical to that of the forward mode AD and is very efficient. 

Mechanically, this PR also separates out the computation of `w` coordinates during decoupling to a separate set of kernels.

### CODE REVIEW

File: tests/test_parameter_interpolation.py [hunk 1]
```diff
@@ -83,4 +84,103 @@ def test_nonbonded(self):
...
+                d4 = jnp.expand_dims(d4, axis=-1)
+                x = jnp.concatenate((x, d4), axis=1)
+
+                qlj_src = params[:len(params)//2]
+                qlj_dst = params[len(params)//2:]
+                qlj = interpolate_params(lamb, qlj_src, qlj_dst)
+                return ref_potential(x, qlj, box, lamb)
+
+
+            for lamb in [0.0, 0.2, 1.0]:
```
Comment (proteneer) [inline tests/test_parameter_interpolation.py:161]: @maxentile usage example here

**Review by maxentile — APPROVED** [review]
This seems like a big step up -- can define protocols flexibly, and automatically get the resulting `du_dl`s required for free energy calculations! This will allow us to run free energy calculations using arbitrary user-specified functions for how to vary distance offsets and nonbonded parameters throughout a protocol.

File: tests/test_parameter_interpolation.py [hunk 2]
```diff
@@ -83,4 +84,104 @@ def test_nonbonded(self):
...
+
+            def u_reference(x, params, box, lamb):
+                d4 = cutoff*(lambda_plane_idxs + lambda_offset_idxs*transform_w(lamb))
+                d4 = jnp.expand_dims(d4, axis=-1)
+                x = jnp.concatenate((x, d4), axis=1)
+
+                qlj_src = params[:len(params)//2]
+                qlj_dst = params[len(params)//2:]
+                qlj = interpolate_params(lamb, qlj_src, qlj_dst)
+                return ref_potential(x, qlj, box, lamb)
```
[the remark this answers is not in the document]
Comment (proteneer) [reply tests/test_parameter_interpolation.py:159]: willfix

File: tests/test_parameter_interpolation.py [hunk 3]
```diff
@@ -83,4 +84,104 @@ def test_nonbonded(self):
...
+
+                    qlj = np.concatenate([qlj_src, qlj_dst])
+
+                    print("lambda", lamb, "cutoff", cutoff, "precision", precision, "xshape", coords.shape)
+
+                    args = copy.deepcopy(test_potential.args)
+                    args.append("lambda*lambda") # transform q
+                    args.append("sin(lambda*PI/2)") # transform sigma
+                    args.append("cos(lambda*PI/2)") # transform epsilon
+                    args.append("1-lambda*lambda") # transform w
```
[the remark this answers is not in the document]
Comment (proteneer) [reply tests/test_parameter_interpolation.py:172]: > Are there any other global constants that can be referenced in these expressions aside from PI?

Right now no, but you can always just hard code in the constants. It woudln't hard to add in other constants though.

> Is it possible to define and reuse variables here? (Something like 0.1 * x + 0.2 * x*x + 0.3 * x*x*x ; x=(1 - lambda*lambda);)

Yes absolutely, but right now the C++ code is written as `return CUSTOM_EXPRESSION`; to support proper branching (eg. if statements beyond simple ternary operators). Currently it's similar to lambda expressions, but  we can definitely relax this without difficulty.

> Is it possible to reference per-atom attributes in these expressions?

Currently no, but if these are forcefield independent attributes, we may be able to support them without too much difficulty. We probably won't be able to do derivatives for the per-atom attributes though, since the forward-mode AD/CSD is only efficient for `R^1->R^N`.

> Is there a practical limit on the length of these expressions?

Anything that doesn't break the compiler :)

File: tests/test_parameter_interpolation.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (proteneer) [reply tests/test_parameter_interpolation.py:159]: will fix

File: timemachine/cpp/src/kernels/k_lambda_transformer_jit.cuh [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+    if(atom_i_idx >= N) {
+        return;
+    }
+
+    int lambda_offset_i = atom_i_idx < N ? lambda_offset_idxs[atom_i_idx] : 0;
+    int lambda_plane_i = atom_i_idx < N ? lambda_plane_idxs[atom_i_idx] : 0;
+
+    double f_lambda = transform_lambda_w(lambda);
+
+    double step = 1e-7;
```
[the remark this answers is not in the document]
Comment (proteneer) [reply timemachine/cpp/src/kernels/k_lambda_transformer_jit.cuh:48]: Yeah you can drop it down to something like `1e-100` if desired. 1e-7 is reasonable enough in the context of fixed point accumulation.

File: timemachine/cpp/src/nonbonded.hpp [hunk 5]
```diff
@@ -78,6 +82,10 @@ class Nonbonded : public Potential {
         cudaStream_t stream
     );
 
+    jitify::KernelInstantiation compute_w_coords_instance_;
```
[the remark this answers is not in the document]
Comment (proteneer) [reply timemachine/cpp/src/nonbonded.hpp:85]: Yep - can of worms has been opened.

File: timemachine/potentials/nonbonded.py [hunk 6]
```diff
@@ -127,10 +127,11 @@ def nonbonded_v3(
     cutoff,
     lambda_plane_idxs,
     lambda_offset_idxs):
-    
+
     N = conf.shape[0]
 
-    conf = convert_to_4d(conf, lamb, lambda_plane_idxs, lambda_offset_idxs, cutoff)
+    if conf.shape[-1] == 3:
```
[the remark this answers is not in the document]
Comment (proteneer) [reply timemachine/potentials/nonbonded.py:133]: Correct! I had to support being the use case where the 4D coordinates were generated outside of this function in the interpolation tests.

File: timemachine/cpp/src/gpu_utils.cuh [hunk 7]
```diff
@@ -36,6 +38,16 @@ inline void curandAssert(curandStatus_t code, const char *file, int line, bool a
...
 
+#define NVRTC_SAFE_CALL(x)                                        \
+  do {                                                            \
+    nvrtcResult result = x;                                       \
+    if (result != NVRTC_SUCCESS) {                                \
+      std::cerr << "\nerror: " #x " failed with error "           \
+                << nvrtcGetErrorString(result) << '\n';           \
+      exit(1);                                                    \
+    }                                                             \
+  } while(0)
```
Comment (badisa) [inline timemachine/cpp/src/gpu_utils.cuh:49]: Is this do while required?

File: timemachine/cpp/src/nonbonded.cu [hunk 8]
```diff
@@ -340,15 +352,18 @@ void Nonbonded<RealType, Interpolated>::execute_device(
...
+        CUresult result = compute_permute_interpolated_.configure(dimGrid, 32, 0, stream)
+        .launch(
             lambda,
             N,
             d_perm_,
             d_p,
             d_sorted_p_,
             d_sorted_dp_dl_
         );
-        gpuErrchk(cudaPeekAtLastError());
```
Comment (badisa) [inline timemachine/cpp/src/nonbonded.cu:351]: This is called on ln 398 after a JIT call, do we still need this?

**Review by badisa — APPROVED** [review]

File: timemachine/cpp/src/gpu_utils.cuh (hunk 7 above)
Comment (proteneer) [reply timemachine/cpp/src/gpu_utils.cuh:49]: It's not - but this macro isn't used, so I will remove it.

File: timemachine/cpp/src/nonbonded.cu (hunk 8 above)
Comment (proteneer) [reply timemachine/cpp/src/nonbonded.cu:351]: willfix
